python/ray/train/v2/_internal/state/state_actor.py

```diff
@@ -1,47 +1,165 @@
+import copy
 import logging
 import os
 import threading
-from collections import defaultdict
+import time
+from collections import OrderedDict, defaultdict
 from typing import Dict, Optional
 
 import ray
+from ray._private import ray_constants
 from ray._private.event.export_event_logger import (
     EventLogType,
     check_export_api_enabled,
     get_export_event_logger,
 )
 from ray.actor import ActorHandle
-from ray.train.v2._internal.state.schema import TrainRun, TrainRunAttempt
+from ray.train.v2._internal.constants import (
+    CONTROLLERS_TO_RECONCILE_PER_CHECK,
+    DEFAULT_ENABLE_STATE_ACTOR_POLLING,
+    DEFAULT_STATE_ACTOR_POLL_INTERVAL_S,
+    ENABLE_STATE_ACTOR_POLLING_ENV_VAR,
+    GET_ACTOR_TIMEOUT_S,
+    STATE_ACTOR_POLL_INTERVAL_S_ENV_VAR,
+)
+from ray.train.v2._internal.state.schema import (
+    TrainRun,
+    TrainRunAttempt,
+)
+from ray.train.v2._internal.state.util import (
+    update_train_run_aborted,
+    update_train_run_attempt_aborted,
+)
+from ray.train.v2._internal.util import time_monotonic
+from ray.util.state import get_actor
 
 logger = logging.getLogger(__name__)
 
 
 class TrainStateActor:
-    def __init__(self):
+    def __init__(
+        self,
+        enable_state_actor_polling: bool,
+        poll_interval_s: float,
+        get_actor_timeout_s: int = GET_ACTOR_TIMEOUT_S,
+        controllers_to_poll_per_iteration: int = CONTROLLERS_TO_RECONCILE_PER_CHECK,
+    ):
         # NOTE: All runs and attempts are stored in memory.
         # This may be a memory issue for large runs.
-        self._runs: Dict[str, TrainRun] = {}
+        # TODO: consider cleaning up runs over time.
+        self._runs: Dict[str, TrainRun] = OrderedDict()
         # {run_id: {attempt_id: TrainRunAttempt}}
-        self._run_attempts: Dict[str, Dict[str, TrainRunAttempt]] = defaultdict(dict)
+        self._run_attempts: Dict[str, OrderedDict[str, TrainRunAttempt]] = defaultdict(
+            OrderedDict
+        )
         (
             self._export_logger,
             self._is_train_run_export_api_enabled,
             self._is_train_run_attempt_export_api_enabled,
         ) = self._init_export_logger()
 
+        # TODO: consider row level locking if loop takes too long.
+        self._runs_lock = threading.RLock()
+        self._run_attempts_lock = threading.RLock()
+
+        # Set env vars related to polling the controller.
+        if enable_state_actor_polling:
+            self._poll_interval_s = poll_interval_s
+            self._controllers_to_poll_per_iteration = controllers_to_poll_per_iteration
+            self._get_actor_timeout_s = get_actor_timeout_s
+            self._start_run_state_reconciliation_thread()
+
+    def _abort_live_runs_with_dead_controllers(
+        self, last_poll_run_id: Optional[str]
```

Comment:
might be simpler if we just keep the idx as an instance variable

Reply:
I used `last_poll_run_id` instead of `last_poll_index` in case we delete old runs in the future.

I didn't create an instance variable because this is only applicable if `enable_state_actor_polling` and I wanted to be consistent with the feedback here: https://github.com/ray-project/ray/pull/53818#discussion_r2162655416